pkg/proxy/ipvs/proxier.go

```diff
@@ -1688,7 +1765,15 @@ func (proxier *Proxier) syncEndpoint(svcPortName proxy.ServicePortName, onlyNode
 		curEndpoints.Insert(des.String())
 	}
 
-	for _, epInfo := range proxier.endpointsMap[svcPortName] {
+	endpoints := proxier.endpointsMap[svcPortName]
+
+	if !onlyNodeLocalEndpoints && utilfeature.DefaultFeatureGate.Enabled(kubefeatures.TopologyAwareServiceRouting) {
```

Comment:
Why the `!onlyNodeLocalEndpoints` check here? Don't think I saw it in the iptables proxy.

Reply:
In other words, if I set `externalTrafficPolicy=Local` and I set topologyKeys that expand outside of nodes, which should take precedent? My first instinct is to assume `topologyKeys` takes precedent over `externalTrafficPolicy` but this probably needs more thought?

Reply:
The whole idea is to use local endpoints when traffic comes from external sources, and use topologically-best set of endpoints for others.

The reason why iptables removed `!svcInfo.OnlyNodeLocalEndpoints()` check is this:
https://github.com/kubernetes/kubernetes/pull/72046#discussion_r315320236
We need to track full set of endpoints, to prevent local endpoints from being filtered out, and filter the topologically-best set of endpoints for non-external sources traffic, the two groups of endpoints was been filtered in one loop, and the ipvs proxier is differrent, it's been filtered in different place by calling `syncEndpoint()` with `onlyNodeLocalEndpoints` indicating whether use only local endpoints, if only use local endpoint (external sources traffic), we don't need to filter the topologically-best set of endpoints, so the `!onlyNodeLocalEndpoints` check was used in ipvs proxier.

Reply:
> the two groups of endpoints was been filtered in one loop, and the ipvs proxier is differrent, it's been filtered in different place by calling syncEndpoint()

Gotcha, was missing this context. Thanks!

Reply:
> We need to track full set of endpoints, to prevent local endpoints from being filtered out, and filter the topologically-best set of endpoints for non-external sources traffic

What about traffic from external sources but with `externalTrafficPolicy=Cluster`? Using topology keys makes sense to me but not exactly consistent with what we are doing with Local? If I set `topologyKeys: [kubernetes.io/hostname]` but `externalTrafficPolicy: Cluster`, this would be the same as `externalTrafficPolicy: Local`, is that expected?

Reply:
Almost seems like there should be an externalTraficPolicy: [] where if its an array, its parsed as a service topology? Maybe there is a reason to have a different topology when traffic is local vs when its external, like maybe taking one layer off of the topology for external?

Reply:
> What about traffic from external sources but with `externalTrafficPolicy=Cluster`? Using topology keys makes sense to me but not exactly consistent with what we are doing with Local? If I set `topologyKeys: [kubernetes.io/hostname]` but `externalTrafficPolicy: Cluster`, this would be the same as `externalTrafficPolicy: Local`, is that expected?

`externalTrafficPolicy:Local`  is used to get the real source IP of external traffic which service topology dose not care about that, so if we set `topologyKeys: ["kubernetes.io/hostname"]` but `externalTrafficPolicy: Cluster`, and traffic from external sources, the endpoints is expected to be the same as `externalTrafficPolicy: Local`, but no masquerade for external traffic (to get the real source IP) if `externalTrafficPolicy: Local`. 
And so it is expected that using topology keys not exactly consistent with what we are doing with Local when `externalTrafficPolicy=Cluster` and traffic from external sources.

Reply:
externalTrafficPolicy is a distinct knob.  If we respect topology for external traffic, then we need to influence which nodes the external LB delivers to.

Concretely:

We can use something like a health check per node, per service that passes if and only if there's an endpoint for that Service on this node.  The external LBs can then dynamically vary the set of nodes they deliver to, ensuring that only nodes with backend get traffic. ... Local policy works.

If we also wanted to do this for Zone, we need each node to know the set of pods on every other node in that zone to do the same thing.  This is clearly not feasible, so now we need a controller and/or API that indicates per-service, which zones it is present in.  

While this may actually be what we want to do long term, we're not doing it now for Alpha. (and we have not really scoped a solution to it at all)

Reply:
Yes, we don't consider this for Alpha, let's consider more in the future.